pallets/gear-builtin-actor/rpc/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! RPC interface for the gear module.
+
+use jsonrpsee::{
+    core::{Error as JsonRpseeError, RpcResult},
+    proc_macros::rpc,
+    types::error::{CallError, ErrorObject},
+};
+pub use pallet_gear_builtin_actor_rpc_runtime_api::GearBuiltinActorApi as GearBuiltinActorRuntimeApi;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_core::H256;
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+
+#[rpc(server)]
+pub trait GearBuiltinActorApi<BlockHash, ResponseType> {
+    #[method(name = "builtinActor_generateActorId")]
+    fn generate_actor_id(&self, builtin_id: u64) -> RpcResult<ResponseType>;
+}
+
+/// Provides RPC methods to query token economics related data.
+pub struct GearBuiltinActor<C, P> {
+    /// Shared reference to the client.
+    client: Arc<C>,
+    _marker: std::marker::PhantomData<P>,
+}
+
+impl<C, P> GearBuiltinActor<C, P> {
+    /// Creates a new instance of the GearBuiltinActor Rpc helper.
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error type of this RPC api.
+pub enum Error {
+    /// The transaction was not decodable.
+    DecodeError,
+    /// The call to runtime failed.
+    RuntimeError,
+}
+
+impl From<Error> for i32 {
+    fn from(e: Error) -> i32 {
+        match e {
+            Error::RuntimeError => 1,
+            Error::DecodeError => 2,
+        }
+    }
+}
+
+impl<C, Block> GearBuiltinActorApiServer<<Block as BlockT>::Hash, H256>
+    for GearBuiltinActor<C, Block>
+where
+    Block: BlockT,
+    C: 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+    C::Api: GearBuiltinActorRuntimeApi<Block>,
+{
+    fn generate_actor_id(&self, builtin_id: u64) -> RpcResult<H256> {
```

Comment:
We can get away without it, of course, if `ProgramIds` corresponding to builtin actors are published somewhere so people (and programs) can send messages to them.
But if you use an rpc like this one you always have a verified `ProgramId` (up-to-date), without having to worry if the hashing algorithm or seed have changed since the `ProgramIds` were published etc. The last thing a program developer wants is to realize they hardcoded a wrong address of a builtin actor when the program has already been uploaded.